cmd/snap/cmd_routine_console_conf.go

```diff
@@ -0,0 +1,184 @@
+	// check if the client has a Maintenance set, in which case we will
+	// skip waiting for changes, etc. and instead will just retry a
+	// connection to snapd until it works, or until we are rebooted
+	maybeMaintErr := cli.Maintenance()
+	if maybeMaintErr == nil {
+		// at least not a maintenance error
+		return false, nil
+	}
+	maintErr, ok := maybeMaintErr.(*client.Error)
+	if !ok {
+		// if cli.Maintenance() didn't return a client.Error we have very weird
+		// problems
+		return true, fmt.Errorf("internal error: client.Maintenance() didn't return a client.Error")
+	}
+
+	if maintErr.Kind == client.ErrorKindDaemonRestart {
+		// then we need to wait for snapd to restart, so keep trying
+		// the console-conf-start endpoint until it works
+		fmt.Fprintf(Stderr, "Snapd is reloading, please wait...\n")
+
+		for {
+			err := retryFunc()
+			if err == nil {
+				// worked, we got a connection and can continue
+				return true, nil
+			}
+
+			// we know that snapd isn't available because it is in
+			// maintenance so we don't gain anything by hitting it
+			// more frequently except for perhaps a quicker latency
+			// for the user when it comes back, but it will be busy
+			// doing things when it starts up anyways so it won't be
+			// able to respond immediately
+			time.Sleep(2 * time.Second)
+		}
+
+	} else if maintErr.Kind == client.ErrorKindSystemRestart {
+		// system is rebooting, just wait for the reboot
+		fmt.Fprintf(Stderr, "System is rebooting, please wait for reboot...\n")
+		time.Sleep(10 * time.Minute)
+		// if we didn't reboot after 10 minutes something's probably broken
+		return true, fmt.Errorf("system didn't reboot after 10 minutes even though snapd daemon is in maintenance")
+	}
+	// other kinds of maintenance error kinds are unhandled, but we still inform
+	// the caller that there was a indeed some kind of maintenance error
+
+	return true, nil
+}
+
+func (x *cmdRoutineConsoleConfStart) Execute(args []string) error {
+	chgs, err := x.client.ConsoleConfStart()
+	if err != nil {
+		wasMaintenance, handlingErr := maybeHandleMaintenance(x.client, func() error {
+			// closure over the chgs variable
+			var startErr error
+			chgs, startErr = x.client.ConsoleConfStart()
+			return startErr
+		})
+		if !wasMaintenance {
+			// some other non-maintenance error hit us
+			return err
+		}
+		if handlingErr != nil {
+			// a maintenance error hit us but we couldn't handle it
+			return fmt.Errorf("while handling maintenance error: %v", handlingErr)
+		}
+
+		// else we handled the maintenance properly and we should have a set of
+		// changes to wait/watch for in the chgs variable
+	}
+
+	// we now handled any on-going maintenance up to the point where we at least
+	// know about any on-going auto-refresh changes, or we handled any
+	// maintenance and there are no more auto-refresh changes
+
+	refreshMsgPrinted := false
+
+	// wait for all the changes that were returned
```

Comment:
my original idea was that much more simply we would poll the original api until it returned that nothing was in progress

Reply:
that would work, it may produce some unnecessary noise by changing the refresh.hold every time we go to change `refresh.hold`, but I suppose that's a feature as it does guarantee that whenever `snap routine console-conf-start` eventually returns, refreshes will be held for exactly 20 minutes after that point, not necessarily for 20 minutes after the interaction started. This is useful for slow internet where downloading your kernel snap update may take more than 20 minutes for example.

I'll make this change

Reply:
done